### ISSUE TITLE
bug and false-positive: pylint blows up with implicit namespaces submodules with relative imports in src-layout in parallel mode

### ISSUE THREAD

**Issue by webknjaz** [pylint-dev/pylint#10147]
### Bug description

```python
$sbj. It either tracebacks or produces false-positives depending on the settings + CLI args. I faced this when replacing a try-except ImportError with just an import in https://github.com/sphinx-contrib/sphinxcontrib-towncrier.git that has both an implicit namespace and an src layout.

When I tried adding `source-roots`, it was giving me more import-related errors which confused me and lead down the wrong rabbit hole. At some point, I realized that `source-roots` does work (one of the complications was the lack of a correct usage example in the docs!), and proceeded bisecting other aspects of the project.

Eventually, I ended up with a reproducer with no deps or config. It turned out that when `--jobs` is anything other than `1`, pylint tracebacks somewhere in `concurrent.futures`. And it seems like depending on what path is passed via CLI, it may either dump that trace or show some import rule violation that is untrue.

See the repro with the console logs below.
```

### Configuration

```ini

```

### Command used

```shell
pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/

pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/relative_module.py

pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/__init__.py

pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/another_relative_module.py

pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/
```

### Pylint output

```python
$ ls -RhalF
.:
total 44K
drwxr-xr-x  3 wk wk  103 Dec 22 05:20  ./
drwxr-xr-x 63 wk wk 8.0K Dec 22 04:49  ../
drwxr-xr-x  3 wk wk   16 Dec 22 04:49  src/

./src:
total 0
drwxr-xr-x 3 wk wk  16 Dec 22 04:49 ./
drwxr-xr-x 3 wk wk 103 Dec 22 05:20 ../
drwxr-xr-x 3 wk wk  24 Dec 22 05:08 ns/

./src/ns:
total 0
drwxr-xr-x 3 wk wk 24 Dec 22 05:08 ./
drwxr-xr-x 3 wk wk 16 Dec 22 04:49 ../
drwxr-xr-x 2 wk wk 85 Dec 22 04:55 concurrent/

./src/ns/concurrent:
total 12K
drwxr-xr-x 2 wk wk 85 Dec 22 04:55 ./
drwxr-xr-x 3 wk wk 24 Dec 22 05:08 ../
-rw-r--r-- 1 wk wk 15 Dec 22 04:54 another_relative_module.py
-rw-r--r-- 1 wk wk 74 Dec 22 04:55 __init__.py
-rw-r--r-- 1 wk wk 41 Dec 22 04:55 relative_module.py

$ cat src/ns/concurrent/__init__.py
from concurrent.futures import Executor

from .relative_module import fun

$ cat src/ns/concurrent/relative_module.py 
from .another_relative_module import fun

$ cat src/ns/concurrent/another_relative_module.py 
def fun(): ...

$ python3.12 -Im venv some-venv-py312

$ some-venv-py312/bin/python -Im pip install pylint
Collecting pylint
  Using cached pylint-3.3.2-py3-none-any.whl.metadata (12 kB)
Collecting dill>=0.3.6 (from pylint)
  Using cached dill-0.3.9-py3-none-any.whl.metadata (10 kB)
Collecting platformdirs>=2.2.0 (from pylint)
  Using cached platformdirs-4.3.6-py3-none-any.whl.metadata (11 kB)
Collecting astroid<=3.4.0-dev0,>=3.3.5 (from pylint)
  Using cached astroid-3.3.7-py3-none-any.whl.metadata (4.5 kB)
Collecting isort!=5.13.0,<6,>=4.2.5 (from pylint)
  Using cached isort-5.13.2-py3-none-any.whl.metadata (12 kB)
Collecting mccabe<0.8,>=0.6 (from pylint)
  Using cached mccabe-0.7.0-py2.py3-none-any.whl.metadata (5.0 kB)
Collecting tomlkit>=0.10.1 (from pylint)
  Using cached tomlkit-0.13.2-py3-none-any.whl.metadata (2.7 kB)
Using cached pylint-3.3.2-py3-none-any.whl (521 kB)
Using cached astroid-3.3.7-py3-none-any.whl (275 kB)
Using cached dill-0.3.9-py3-none-any.whl (119 kB)
Using cached isort-5.13.2-py3-none-any.whl (92 kB)
Using cached mccabe-0.7.0-py2.py3-none-any.whl (7.3 kB)
Using cached platformdirs-4.3.6-py3-none-any.whl (18 kB)
Using cached tomlkit-0.13.2-py3-none-any.whl (37 kB)
Installing collected packages: tomlkit, platformdirs, mccabe, isort, dill, astroid, pylint
Successfully installed astroid-3.3.7 dill-0.3.9 isort-5.13.2 mccabe-0.7.0 platformdirs-4.3.6 pylint-3.3.2 tomlkit-0.13.2

$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=1 src/
No config file found, using default configuration

----------------------------------------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.00/10, +2.00)
Checked 3 files, skipped 3 files


$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=0 src/
No config file found, using default configuration
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__main__.py", line 10, in <module>
    pylint.run_pylint()
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/run.py", line 215, in __init__
    linter.check(args)
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 679, in check
    check_parallel(
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/parallel.py", line 162, in check_parallel
    ) in executor.map(_worker_check_single_file, files):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 851, in map
    results = super().map(partial(_process_chunk, fn),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 608, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
                                             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 198, in _get_chunks
    chunk = tuple(itertools.islice(it, chunksize))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 874, in _iterate_file_descrs
    for descr in self._expand_files(files_or_modules).values():
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 883, in _expand_files
    result, errors = expand_modules(
                     ^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 158, in expand_modules
    modpath = _modpath_from_file(
              ^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 22, in _modpath_from_file
    return modutils.modpath_from_file_with_callback(  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/astroid/modutils.py", line 294, in modpath_from_file_with_callback
    raise ImportError(
ImportError: Unable to find module for src/ns/concurrent/another_relative_module.py in /usr/lib/python312.zip, 
/usr/lib/python3.12, 
/usr/lib/python3.12/lib-dynload, 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages

$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/
No config file found, using default configuration
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__main__.py", line 10, in <module>
    pylint.run_pylint()
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/run.py", line 215, in __init__
    linter.check(args)
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 679, in check
    check_parallel(
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/parallel.py", line 162, in check_parallel
    ) in executor.map(_worker_check_single_file, files):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 851, in map
    results = super().map(partial(_process_chunk, fn),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 608, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
                                             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 198, in _get_chunks
    chunk = tuple(itertools.islice(it, chunksize))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 874, in _iterate_file_descrs
    for descr in self._expand_files(files_or_modules).values():
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 883, in _expand_files
    result, errors = expand_modules(
                     ^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 158, in expand_modules
    modpath = _modpath_from_file(
              ^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 22, in _modpath_from_file
    return modutils.modpath_from_file_with_callback(  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/astroid/modutils.py", line 294, in modpath_from_file_with_callback
    raise ImportError(
ImportError: Unable to find module for src/ns/concurrent/another_relative_module.py in /usr/lib/python312.zip, 
/usr/lib/python3.12, 
/usr/lib/python3.12/lib-dynload, 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages

$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent 
No config file found, using default configuration
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__main__.py", line 10, in <module>
    pylint.run_pylint()
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/run.py", line 215, in __init__
    linter.check(args)
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 679, in check
    check_parallel(
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/parallel.py", line 162, in check_parallel
    ) in executor.map(_worker_check_single_file, files):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 851, in map
    results = super().map(partial(_process_chunk, fn),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 608, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
                                             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 198, in _get_chunks
    chunk = tuple(itertools.islice(it, chunksize))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 874, in _iterate_file_descrs
    for descr in self._expand_files(files_or_modules).values():
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 883, in _expand_files
    result, errors = expand_modules(
                     ^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 158, in expand_modules
    modpath = _modpath_from_file(
              ^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 22, in _modpath_from_file
    return modutils.modpath_from_file_with_callback(  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/astroid/modutils.py", line 294, in modpath_from_file_with_callback
    raise ImportError(
ImportError: Unable to find module for src/ns/concurrent/another_relative_module.py in /usr/lib/python312.zip, 
/usr/lib/python3.12, 
/usr/lib/python3.12/lib-dynload, 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages

$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/another_relative_module.py 
No config file found, using default configuration

-----------------------------------------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
Checked 1 files, skipped 1 files


$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/__init__.py 
No config file found, using default configuration
************* Module __init__
src/ns/concurrent/__init__.py:3:0: E0401: Unable to import '__init__.relative_module' (import-error)

---------------------------------------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
Checked 1 files, skipped 1 files


$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/relative_module.py 
No config file found, using default configuration
************* Module relative_module
src/ns/concurrent/relative_module.py:1:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)

--------------------------------------------------------------------
Your code has been rated at 0.00/10
Checked 1 files, skipped 1 files


$ some-venv-py312/bin/python -Im pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/                  
No config file found, using default configuration
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__main__.py", line 10, in <module>
    pylint.run_pylint()
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/run.py", line 215, in __init__
    linter.check(args)
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 679, in check
    check_parallel(
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/parallel.py", line 162, in check_parallel
    ) in executor.map(_worker_check_single_file, files):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 851, in map
    results = super().map(partial(_process_chunk, fn),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 608, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
                                             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 198, in _get_chunks
    chunk = tuple(itertools.islice(it, chunksize))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 874, in _iterate_file_descrs
    for descr in self._expand_files(files_or_modules).values():
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 883, in _expand_files
    result, errors = expand_modules(
                     ^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 158, in expand_modules
    modpath = _modpath_from_file(
              ^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 22, in _modpath_from_file
    return modutils.modpath_from_file_with_callback(  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/astroid/modutils.py", line 294, in modpath_from_file_with_callback
    raise ImportError(
ImportError: Unable to find module for src/ns/concurrent/another_relative_module.py in /usr/lib/python312.zip, 
/usr/lib/python3.12, 
/usr/lib/python3.12/lib-dynload, 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages

➜ some-venv-py312/bin/python -Im pip install 'astroid @ git+https://github.com/pylint-dev/astroid.git' 'pylint @ git+https://github.com/pylint-dev/pylint.git'
Collecting astroid@ git+https://github.com/pylint-dev/astroid.git
  Cloning https://github.com/pylint-dev/astroid.git to /tmp/pip-install-p0wd_ffk/astroid_2dbb18c524b0442d8ef658bf73fbf2d1
  Running command git clone --filter=blob:none --quiet https://github.com/pylint-dev/astroid.git /tmp/pip-install-p0wd_ffk/astroid_2dbb18c524b0442d8ef658bf73fbf2d1
  Resolved https://github.com/pylint-dev/astroid.git to commit 30ddb250a4c15aadf5d397fa42a96a2ac108c8d8
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting pylint@ git+https://github.com/pylint-dev/pylint.git
  Cloning https://github.com/pylint-dev/pylint.git to /tmp/pip-install-p0wd_ffk/pylint_77f85aaa0c9c491e85b3679b3d637a5f
  Running command git clone --filter=blob:none --quiet https://github.com/pylint-dev/pylint.git /tmp/pip-install-p0wd_ffk/pylint_77f85aaa0c9c491e85b3679b3d637a5f
  Resolved https://github.com/pylint-dev/pylint.git to commit be1968e6aa59c8de19a778c280b192e53471a96f
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: dill>=0.3.6 in ./some-venv-py312/lib/python3.12/site-packages (from pylint@ git+https://github.com/pylint-dev/pylint.git) (0.3.9)
Requirement already satisfied: platformdirs>=2.2.0 in ./some-venv-py312/lib/python3.12/site-packages (from pylint@ git+https://github.com/pylint-dev/pylint.git) (4.3.6)
Requirement already satisfied: isort!=5.13.0,<6,>=4.2.5 in ./some-venv-py312/lib/python3.12/site-packages (from pylint@ git+https://github.com/pylint-dev/pylint.git) (5.13.2)
Requirement already satisfied: mccabe<0.8,>=0.6 in ./some-venv-py312/lib/python3.12/site-packages (from pylint@ git+https://github.com/pylint-dev/pylint.git) (0.7.0)
Requirement already satisfied: tomlkit>=0.10.1 in ./some-venv-py312/lib/python3.12/site-packages (from pylint@ git+https://github.com/pylint-dev/pylint.git) (0.13.2)
Building wheels for collected packages: astroid, pylint
  Building wheel for astroid (pyproject.toml) ... done
  Created wheel for astroid: filename=astroid-4.0.0.dev0-py3-none-any.whl size=275991 sha256=79a792879efe3abbd977258a2c851d750004182961472a970701cea1dceb983b
  Stored in directory: /tmp/pip-ephem-wheel-cache-cx4l0ij1/wheels/ee/2d/6e/311cb12113446220dc04dc7a7284d966fd348a5fa7f7948766
  Building wheel for pylint (pyproject.toml) ... done
  Created wheel for pylint: filename=pylint-4.0.0.dev0-py3-none-any.whl size=522853 sha256=9326efdf2ad5219ddeb659b9269616afc9f3796762e11f9651ba8de512fa88bd
  Stored in directory: /tmp/pip-ephem-wheel-cache-cx4l0ij1/wheels/61/54/d0/bfb7060be86de9deff91d06ca91aa55805e79a8c9e4fa59f6d
Successfully built astroid pylint
Installing collected packages: astroid, pylint
  Attempting uninstall: astroid
    Found existing installation: astroid 3.3.7
    Uninstalling astroid-3.3.7:
      Successfully uninstalled astroid-3.3.7
  Attempting uninstall: pylint
    Found existing installation: pylint 3.3.2
    Uninstalling pylint-3.3.2:
      Successfully uninstalled pylint-3.3.2
Successfully installed astroid-4.0.0.dev0 pylint-4.0.0.dev0

src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout took 8.9s …
➜ some-venv-py312/bin/python -m pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 src/ns/concurrent/
No config file found, using default configuration
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__main__.py", line 10, in <module>
    pylint.run_pylint()
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/run.py", line 215, in __init__
    linter.check(args)
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 679, in check
    check_parallel(
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/parallel.py", line 162, in check_parallel
    ) in executor.map(_worker_check_single_file, files):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 851, in map
    results = super().map(partial(_process_chunk, fn),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 608, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
                                             ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 198, in _get_chunks
    chunk = tuple(itertools.islice(it, chunksize))
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 874, in _iterate_file_descrs
    for descr in self._expand_files(files_or_modules).values():
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/pylinter.py", line 883, in _expand_files
    result, errors = expand_modules(
                     ^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 158, in expand_modules
    modpath = _modpath_from_file(
              ^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/pylint/lint/expand_modules.py", line 22, in _modpath_from_file
    return modutils.modpath_from_file_with_callback(  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages/astroid/modutils.py", line 294, in modpath_from_file_with_callback
    raise ImportError(
ImportError: Unable to find module for src/ns/concurrent/another_relative_module.py in ., 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/src, 
/usr/lib/python312.zip, 
/usr/lib/python3.12, 
/usr/lib/python3.12/lib-dynload, 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages, 
/usr/lib/python312.zip, 
/usr/lib/python3.12, 
/usr/lib/python3.12/lib-dynload, 
~/src/experiments/pylint-repro-relative-import-error-implicit-namespace-src-layout/some-venv-py312/lib/python3.12/site-packages
```

### Expected behavior

No false positives (`import-error`, `no-name-in-module`, `relative-beyond-top-level` seem to be caused by this bug) or tracebacks.

### Pylint version

```shell
$ some-venv-py312/bin/python -Im pylint --version
pylint 3.3.2
astroid 3.3.7
Python 3.12.8 (main, Dec 11 2024, 12:13:45) [GCC 14.2.1 20241116]

$ some-venv-py312/bin/python -Im pip list
Package      Version
------------ -------
astroid      3.3.7
dill         0.3.9
isort        5.13.2
mccabe       0.7.0
pip          24.3.1
platformdirs 4.3.6
pylint       3.3.2
tomlkit      0.13.2
```

### OS / Environment

Only checked under GNU/Linux. It's Ubuntu that GHA provides and my laptop with Gentoo Linux. Python 3.9, 3.10, 3.11, 3.12. I haven't checked others, but I imagine it'd be the same.

### Additional dependencies

```python

```

**Comment by webknjaz** [comment]
I'm also getting a `no-name-in-module` w/o parallelism. Haven't yet figured out the pattern...

**Comment by Julfried** [comment]
As discussed here in #10185 I think the problem is likely the way you call pylint. By setting source roots you tell pylint that the namespace starts inside the folder you specified, so you file paths need to start from there:

```
pylint --verbose --disable=missing-function-docstring,missing-module-docstring,unused-import --output-format=colorized --source-roots=src/ --recursive=y --jobs=2 ns/concurrent/
```

**Comment by webknjaz** [comment]
@Julfried it's important to note that when using the pre-commit framework, there's no way to tell it to make the paths relative to non-repo root.

**Comment by webknjaz** [comment]
Trying that with `pass_filenames: false` does not help:
```console
src/awx_plugins/credentials/plugin.py:10:0: E0611: No name '_types' in module '' (no-name-in-module)
```

**Comment by webknjaz** [comment]
Though, this might be an additional false-positive..

**Comment by Julfried** [comment]
I think I found the issue. Could you try [my branch](https://github.com/Julfried/pylint/tree/parallel) and let me know if that fixes it for you?

**Comment by webknjaz** [comment]
@Julfried thanks, I'll check! Though, reverting https://github.com/ansible/awx-plugins/commit/805ddb27cf8eb3e0599c5e4088e67273e10c4025 on top of https://github.com/ansible/awx-plugins/pull/161 and setting `jobs = 1` still exhibits the issue. So it's unlikely that your patch will affect the last problem I observed. Note that the last error I posted was from this PR (161) and the layout there is `src/awx_plugins/{credentials,inventory}/*.py` within that repo but also there's `src/awx_plugins/interfaces/*` in a dependency that gets installed into site-packages. And so the Python namespace package is spread across a few locations on disk: whatever virtualenv the `pre-commit` framework manages gets `awx_plugins.interfaces` in it, some test deps and pylint itself, and the linted content is in the repo dir in `src/awx_plugins/`, `tests/`, `bin/` and a `.py` file in the repo root. The problems seem to originate from pylint trying to merge the importable modules that need to be in the same namespace but might be installed into different dirs.

**Comment by webknjaz** [comment]
Oh, and I just realized that this project makes use of the `init-hook` that prepends `src` to import paths: https://github.com/ansible/awx-plugins/blob/37a1e756ef01567f2c92bba7476342d3e02f8577/.pylintrc.toml#L64-L70.

**Comment by webknjaz** [comment]
Confirmed: that commit didn't change anything for this problem.

**Comment by webknjaz** [comment]
With some breakpoints, I discovered that this is where the error is being reported: https://github.com/pylint-dev/pylint/blob/24a69661e1869c78a417649de88cf3976ceac669/pylint/checkers/variables.py#L3204-L3206. And at this point, `module` seems to be https://github.com/pylint-dev/astroid/blob/c7908a16d1cf0a85de6f1723eefe6b0c8cc1115e/astroid/brain/brain_builtin_inference.py#L46-L86:

```pycon
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3190)_check_module_attrs()
-> breakpoint()
(Pdb) pp module_names
[]
(Pdb) c
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3190)_check_module_attrs()
-> breakpoint()
(Pdb) pp module_names
['_types']
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3191)_check_module_attrs()
-> while module_names:
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3192)_check_module_attrs()
-> name = module_names.pop(0)
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3193)_check_module_attrs()
-> if name == "__dict__":
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3196)_check_module_attrs()
-> try:
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(3197)_check_module_attrs()
-> module = module.getattr(name)[0]
(Pdb) pp module
<Module l.0 at 0x7fbe98fcccd0>
(Pdb) pp module.getattr
<bound method Module.getattr of <Module l.0 at 0x7fbe98fcccd0>>
(Pdb) pp module.getattr(name)
*** astroid.exceptions.AttributeInferenceError: '_types' not found on <Module l.0 at 0x7fbe98fcccd0>.
(Pdb) pp module.as_string()
('\n'
 '\n'
 'class whatever(object):\n'
 '    \n'
 '    def join(self, iterable):\n'
 "        return b''\n"
 '    \n'
 '    def replace(self, old, new, count=None):\n'
 "        return b''\n"
 '    \n'
 "    def decode(self, encoding='ascii', errors=None):\n"
 "        return ''\n"
 '    \n'
 '    def capitalize(self):\n'
 "        return b''\n"
 '    \n'
 '    def title(self):\n'
 "        return b''\n"
 '    \n'
 '    def lower(self):\n'
 "        return b''\n"
 '    \n'
 '    def upper(self):\n'
 "        return b''\n"
 '    \n'
 '    def swapcase(self):\n'
 "        return b''\n"
 '    \n'
 '    def index(self, sub, start=None, end=None):\n'
 '        return 0\n'
 '    \n'
 '    def find(self, sub, start=None, end=None):\n'
 '        return 0\n'
 '    \n'
 '    def count(self, sub, start=None, end=None):\n'
 '        return 0\n'
 '    \n'
 '    def strip(self, chars=None):\n'
 "        return b''\n"
 '    \n'
 '    def lstrip(self, chars=None):\n'
 "        return b''\n"
 '    \n'
 '    def rstrip(self, chars=None):\n'
 "        return b''\n"
 '    \n'
 '    def rjust(self, width, fillchar=None):\n'
 "        return b''\n"
 '    \n'
 '    def center(self, width, fillchar=None):\n'
 "        return b''\n"
 '    \n'
 '    def ljust(self, width, fillchar=None):\n'
 "        return b''\n"
 '\n'
 '\n')
(Pdb)
```

**Comment by webknjaz** [comment]
Setting a breakpoint at https://github.com/pylint-dev/pylint/blob/24a69661e1869c78a417649de88cf3976ceac669/pylint/checkers/variables.py#L2122, reveals that https://github.com/pylint-dev/pylint/blob/24a69661e1869c78a417649de88cf3976ceac669/pylint/checkers/variables.py#L2135 probably gets confused with an empty string import request:
```pycon
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2125)visit_importfrom()
-> breakpoint()
(Pdb) args
self = Checker 'variables' (responsible for 'E0601', 'E0602', 'E0603', 'E0604', 'E0605', 'E0606', 'E0611', 'W0601', 'W0602', 'W0603', 'W0604', 'W0611', 'W0612', 'W0613', 'W0614', 'W0621', 'W0622', 'W0631', 'W0632', 'E0633', 'W0640', 'W0641', 'W0642', 'E0643', 'W0644')
node = <ImportFrom l.12 at 0x7f5573564640>
(Pdb) pp node.as_string()
'from . import _types'
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2126)visit_importfrom()
-> if not self._analyse_fallback_blocks and utils.is_from_fallback_block(node):
(Pdb) pp self._analyse_fallback_blocks
False
(Pdb) pp node.parent
<Module.aim l.0 at 0x7f5573564510>
(Pdb) pp node.parent.package
False
(Pdb) pp node.parent.path
['~/src/github/ansible/awx-plugins--chore--typing/src/awx_plugins/credentials/aim.py']
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2132)visit_importfrom()
-> if in_type_checking_block(node):
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2134)visit_importfrom()
-> if isinstance(node.parent, nodes.If) and is_sys_guard(node.parent):
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2137)visit_importfrom()
-> name_parts = node.modname.split(".")
(Pdb) pp node.modname
''
(Pdb) pp node.position
None
(Pdb) pp node.modname
''
(Pdb) pp node.parent
<Module.aim l.0 at 0x7f5573564510>
(Pdb) pp node.parent.name
'aim'
(Pdb) pp node.parent.parent
None
(Pdb) pp node.parent.package
False
(Pdb) pp node.modname
''
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2138)visit_importfrom()
-> try:
(Pdb) pp node.do_import_module(name_parts[0])
<Module l.0 at 0x7f55736fccd0>
(Pdb) pp name_parts
['']
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2139)visit_importfrom()
-> module = node.do_import_module(name_parts[0])
(Pdb) pp module
*** NameError: name 'module' is not defined
(Pdb) n
> ~/.cache/pre-commit/repolhbtj5zy/py_env-python3.13/lib/python3.13/site-packages/pylint/checkers/variables.py(2142)visit_importfrom()
-> module = self._check_module_attrs(node, module, name_parts[1:])
(Pdb) pp module
<Module l.0 at 0x7f55736fccd0>
(Pdb) pp module.name
''
(Pdb) pp module.package
False
(Pdb) pp module.parent
None
(Pdb) pp module.path
['<?>']
(Pdb) pp (node, module, name_parts[1:])
(<ImportFrom l.12 at 0x7f5573564640>, <Module l.0 at 0x7f55736fccd0>, [])
(Pdb) pp node.modname
''
(Pdb) 
```

**Comment by Julfried** [comment]
Sounds reasonable, so I guess this is something that should be fixed in astroid? @Pierre-Sassoulas what do you think?

**Comment by Pierre-Sassoulas** [comment]
I didn't have time to look at it in détails but it seems a big part of the analysis is already done here.

**Comment by webknjaz** [comment]
@Pierre-Sassoulas I thought we agreed that this is not related to multiprocessing, no? (it's possible I've got two cases mixed up by now, though)

**Comment by Pierre-Sassoulas** [comment]
Hey @webknjaz, I tried a fix, would you mind checking if the problem reproduce with ``pip install git+https://github.com/pierre-sassoulas/pylint.git@d28b79a41e985d4e3a255773b211f31f225bf401``, please ?

**Comment by webknjaz** [comment]
I'll try but your commit message implies it's still related to parallel mode, why?

**Comment by webknjaz** [comment]
@Pierre-Sassoulas with your commit and `jobs = 1` (plus `no-name-in-module`, `relative-beyond-top-level` dropped from ignores) `pre-commit run pylint --all-files -vv` generates both violations in https://github.com/ansible/awx-plugins.

Though, when I checked https://github.com/sphinx-contrib/sphinxcontrib-towncrier, this commit seems to have fixed its issues with `jobs = 0` (that weren't present w/ `jobs = 1`). So I guess we've got two distinct problems on our hands. Or rather that with `sphinxcontrib-towncrier`, there's no second dependency from the same namespace in the virtualenv (which is likely why it doesn't trigger that last problem I observed), while with `awx-plugins` there is.

**Comment by Julfried** [comment]
@Pierre-Sassoulas Funny enough this is actual a similar fix to what I proposed in here. Added some regression tests yesterday and just wanted to open a PR :)
> I think I found the issue. Could you try [my branch](https://github.com/Julfried/pylint/tree/parallel) and let me know if that fixes it for you?

**Comment by Pierre-Sassoulas** [comment]
Yeah, it's the same fix 😄. Damn. Maybe I should not assume a 1-week-old message is a dead end and actually look at it. I'm looking at the second bug now. Do you have any branch up your sleeve where it's already fixed ? 😉 (Please open a PR @Julfried )

**Comment by Julfried** [comment]
No worries, stuff like this happens. I added the tests in the same branch, leave it up to you which one you want to merge :)
https://github.com/Julfried/pylint/tree/parallel

Ok, will do the PR later👍

**Comment by Pierre-Sassoulas** [comment]
Could you try again with ``pip install git+https://github.com/pierre-sassoulas/pylint.git@963057a5efcecdd9d040faa53f0f7c3b93eaeb8e``, please @webknjaz ?

**Comment by webknjaz** [comment]
@Pierre-Sassoulas in your commit message, you included the implication
> that use init-hook to add src/ to sys.path
> rather than --source-roots.

into your commit message. Upon seeing that, I realized that I had both `source-roots` and `init-hook` in `sphinxcontrib-towncrier` and so I tested that `963057a5efcecdd9d040faa53f0f7c3b93eaeb8e` does fix all the problem aspects I had in that repo, including both ways of setting source roots (one at a time), and with wither of `jobs = 1` or `jobs = 0`. Namely, it fixed `import-error` and `relative-beyond-top-level` in `sphinxcontrib-towncrier`.

But in `awx-plugins`, it seems to have no effect and both `no-name-in-module` + `relative-beyond-top-level` keep being reported when they shouldn't be. Even in a `jobs = 1` mode. So I looked at the regression test you added and realized that this project has relative imports crashing the check, even though the file isn't in an external folder — it's right next to the module where the false-positive happens:
```console
$ grep _types src/awx_plugins/credentials/tss.py
from . import _types
tss_inputs: _types.PluginInputs = {

$ ls -lFh src/awx_plugins/credentials/_types.py           
-rw-r--r-- 1 <user> <user> 946 Mar  4 17:08 src/awx_plugins/credentials/_types.py
```
